agent/agent.go

```diff
@@ -1000,19 +1016,74 @@ func Bicopy(ctx context.Context, c1, c2 io.ReadWriteCloser) {
 	}
 }
 
-// ExpandRelativeHomePath expands the tilde at the beginning of a path to the
-// current user's home directory and returns a full absolute path.
-func ExpandRelativeHomePath(in string) (string, error) {
-	usr, err := user.Current()
+// isQuietLogin checks if the SSH server should perform a quiet login or not.
+//
+// https://github.com/openssh/openssh-portable/blob/25bd659cc72268f2858c5415740c442ee950049f/session.c#L816
+func isQuietLogin(rawCommand string) bool {
+	// We are always quiet unless this is a login shell.
+	if len(rawCommand) != 0 {
+		return true
+	}
+
+	// Best effort, if we can't get the home directory,
+	// we can't lookup .hushlogin.
+	homedir, err := getHomeDir()
+	if err == nil {
+		if _, err := os.Stat(filepath.Join(homedir, ".hushlogin")); err == nil {
+			return true
+		}
+	}
+
+	return false
+}
+
+// showMOTD will output the message of the day from
+// the given filename to dest, if the file exists.
+//
+// https://github.com/openssh/openssh-portable/blob/25bd659cc72268f2858c5415740c442ee950049f/session.c#L784
+func showMOTD(dest io.Writer, filename string) error {
+	if filename == "" {
+		return nil
+	}
+
+	f, err := os.Open(filename)
 	if err != nil {
-		return "", xerrors.Errorf("get current user details: %w", err)
+		if xerrors.Is(err, os.ErrNotExist) {
+			// This is not an error, there simply isn't a MOTD to show.
+			return nil
+		}
+		return xerrors.Errorf("open MOTD: %w", err)
 	}
+	defer f.Close()
 
-	if in == "~" {
-		in = usr.HomeDir
-	} else if strings.HasPrefix(in, "~/") {
-		in = filepath.Join(usr.HomeDir, in[2:])
+	s := bufio.NewScanner(f)
+	for s.Scan() {
+		// Carriage return ensures each line starts
+		// at the beginning of the terminal.
+		_, err = fmt.Fprint(dest, s.Text()+"\r\n")
+		if err != nil {
+			return xerrors.Errorf("write MOTD: %w", err)
+		}
+	}
+	if err := s.Err(); err != nil {
+		return xerrors.Errorf("read MOTD: %w", err)
 	}
 
-	return filepath.Abs(in)
+	return nil
+}
+
+// getHomeDir returns the home directory of the current user, giving
+// priority to the $HOME environment variable.
+func getHomeDir() (string, error) {
+	// First we check the environment.
+	homedir, err := os.UserHomeDir()
+	if err != nil {
```

Comment:
I don’t think it’s worth departing from `if err != nil` here to make it flatter. Typically I try to avoid `if err == nil` checks because they can lead to the error not being handled.

Reply:
My point is to adapt the code to the [return early pattern](https://medium.com/swlh/return-early-pattern-3d18a41bba8).

Reply:
I know the pattern, and I like it, but how would you apply it to this piece of code whilst _not_ doing an `err == nil` check?